// @qijenchen/design-system/dist/patterns/resize-handle/resize-handle.js
import { jsx } from "react/jsx-runtime";
import * as React from "react";
import { cn } from "../../lib/utils.js";
const ResizeHandle = React.forwardRef(
  ({
    direction,
    position = "end",
    isResizing,
    disabled,
    showLine = true,
    lineInsetStart,
    lineInsetEnd,
    className,
    ...props
  }, ref) => {
    const isHorizontal = direction === "horizontal";
    const ariaOrientation = isHorizontal ? "vertical" : "horizontal";
    const hitZoneStyle = isHorizontal ? {
      position: "absolute",
      top: 0,
      bottom: 0,
      width: 7,
      ...position === "end" ? { right: 0, marginRight: -3 } : { left: 0, marginLeft: -3 },
      cursor: disabled ? void 0 : "col-resize"
    } : {
      position: "absolute",
      left: 0,
      right: 0,
      height: 7,
      ...position === "end" ? { bottom: 0, marginBottom: -3 } : { top: 0, marginTop: -3 },
      cursor: disabled ? void 0 : "row-resize"
    };
    const lineColorClass = isResizing ? "bg-primary" : disabled ? "bg-divider" : "bg-divider group-hover/resize:bg-[var(--border-hover)]";
    const lineStyle = isHorizontal ? {
      position: "absolute",
      width: 1,
      top: lineInsetStart ?? 0,
      bottom: lineInsetEnd ?? 0,
      ...position === "end" ? { right: 3 } : { left: 3 }
    } : {
      position: "absolute",
      height: 1,
      left: lineInsetStart ?? 0,
      right: lineInsetEnd ?? 0,
      ...position === "end" ? { bottom: 3 } : { top: 3 }
    };
    const { "aria-label": ariaLabel, style: extraStyle, ...restProps } = props;
    return /* @__PURE__ */ jsx(
      "span",
      {
        ref,
        role: disabled ? void 0 : "separator",
        "aria-orientation": disabled ? void 0 : ariaOrientation,
        "aria-label": disabled ? void 0 : ariaLabel,
        style: { ...hitZoneStyle, ...extraStyle },
        className: cn("group/resize", !disabled && "select-none", className),
        ...restProps,
        children: showLine && /* @__PURE__ */ jsx(
          "span",
          {
            "aria-hidden": true,
            className: cn("transition-colors", lineColorClass),
            style: lineStyle
          }
        )
      }
    );
  }
);
ResizeHandle.displayName = "ResizeHandle";
export {
  ResizeHandle
};
