// @qijenchen/design-system/dist/components/DataTable/data-table-interaction-layer.js
import { jsxs, jsx } from "react/jsx-runtime";
import * as React from "react";
function getCellRect(containerEl, cellId) {
  if (!containerEl) return null;
  const cellEl = containerEl.querySelector(`[data-cell-id="${cellId}"]`);
  if (!cellEl) return null;
  const rect = cellEl.getBoundingClientRect();
  return {
    x: rect.x,
    y: rect.y,
    width: rect.width,
    height: rect.height
  };
}
function getCellGeometry(containerEl, cellId) {
  if (!containerEl) return null;
  const cellEl = containerEl.querySelector(`[data-cell-id="${cellId}"]`);
  if (!cellEl) return null;
  const panelEl = cellEl.closest("[data-datatable-panel]");
  if (!panelEl) return null;
  const panel = panelEl.dataset.datatablePanel;
  const cellR = cellEl.getBoundingClientRect();
  const panelR = panelEl.getBoundingClientRect();
  return {
    rect: { x: cellR.x, y: cellR.y, width: cellR.width, height: cellR.height },
    clipRect: { x: panelR.x, y: panelR.y, width: panelR.width, height: panelR.height },
    panel
  };
}
function useReactiveRect(containerRef, enabled) {
  const [, setVersion] = React.useState(0);
  React.useEffect(() => {
    if (!enabled) return;
    let rafId = null;
    const force = () => {
      if (rafId !== null) return;
      rafId = requestAnimationFrame(() => {
        rafId = null;
        setVersion((v) => v + 1);
      });
    };
    window.addEventListener("scroll", force, { capture: true, passive: true });
    window.addEventListener("resize", force, { passive: true });
    const container = containerRef.current;
    let ro = null;
    if (container && typeof ResizeObserver !== "undefined") {
      ro = new ResizeObserver(force);
      ro.observe(container);
    }
    return () => {
      window.removeEventListener("scroll", force, { capture: true });
      window.removeEventListener("resize", force);
      if (ro) ro.disconnect();
      if (rafId !== null) cancelAnimationFrame(rafId);
    };
  }, [enabled, containerRef]);
}
function useHoveredCell(containerRef, enabled) {
  const [hoveredCellId, setHoveredCellId] = React.useState(null);
  React.useEffect(() => {
    if (!enabled) {
      setHoveredCellId(null);
      return;
    }
    const container = containerRef.current;
    if (!container) return;
    const handleMouseOver = (e) => {
      const target = e.target;
      const cellEl = target == null ? void 0 : target.closest("[data-cell-id]");
      const id = (cellEl == null ? void 0 : cellEl.dataset.cellId) ?? null;
      setHoveredCellId(id);
    };
    const handleMouseLeave = () => setHoveredCellId(null);
    container.addEventListener("mouseover", handleMouseOver);
    container.addEventListener("mouseleave", handleMouseLeave);
    return () => {
      container.removeEventListener("mouseover", handleMouseOver);
      container.removeEventListener("mouseleave", handleMouseLeave);
    };
  }, [enabled, containerRef]);
  return hoveredCellId;
}
function DataTableInteractionLayer({
  enabled,
  containerRef,
  cellClickEntersEdit,
  activeEditorCellId,
  activeEditorEnabled = false,
  activeEditorRender,
  selectedCellId = null,
  // rangeCellIds retired 2026-05-10(outer ring 已 retire,range visual 靠 cell-bg `[data-range-cell]`)
  rangeCellIds: _rangeCellIds
}) {
  const hoveredCellId = useHoveredCell(containerRef, enabled);
  useReactiveRect(containerRef, enabled);
  const isEditingHovered = activeEditorCellId != null && activeEditorCellId === hoveredCellId;
  const isSelectedHovered = selectedCellId != null && selectedCellId === hoveredCellId;
  const shouldShowHover = enabled && hoveredCellId != null && !isEditingHovered && !isSelectedHovered && (cellClickEntersEdit ? cellClickEntersEdit(hoveredCellId) : true);
  const hoverGeo = shouldShowHover ? getCellGeometry(containerRef.current, hoveredCellId) : null;
  const activeEditorRect = activeEditorEnabled && activeEditorCellId != null ? getCellRect(containerRef.current, activeEditorCellId) : null;
  const selectedGeo = enabled && selectedCellId != null ? getCellGeometry(containerRef.current, selectedCellId) : null;
  if (!enabled) return null;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "aria-hidden": true,
      style: { position: "fixed", inset: 0, pointerEvents: "none", zIndex: 1 },
      children: [
        hoverGeo && /* @__PURE__ */ jsx(ClipMask, { clipRect: hoverGeo.clipRect, children: /* @__PURE__ */ jsx(CellRingOverlay, { rect: toRelRect(hoverGeo.rect, hoverGeo.clipRect), kind: "hover" }) }),
        selectedGeo && /* @__PURE__ */ jsx(ClipMask, { clipRect: selectedGeo.clipRect, children: /* @__PURE__ */ jsx(
          CellRingOverlay,
          {
            rect: toRelRect(selectedGeo.rect, selectedGeo.clipRect),
            kind: "selected",
            cellId: selectedCellId
          }
        ) }),
        activeEditorRect && /* @__PURE__ */ jsx(ActiveEditorHost, { rect: activeEditorRect, children: activeEditorRender ? activeEditorRender(activeEditorCellId, activeEditorRect) : null })
      ]
    }
  );
}
function toRelRect(cellRect, clipRect) {
  return {
    x: cellRect.x - clipRect.x,
    y: cellRect.y - clipRect.y,
    width: cellRect.width,
    height: cellRect.height
  };
}
const CELL_RING_STYLES = {
  hover: { width: 1, color: "var(--border-hover)", zIndex: 1 },
  selected: { width: 1, color: "var(--primary)", zIndex: 2 }
  // future kinds(per codex Q6 outline,加 entry 即可擴展):
  // focus:  { width: 2, color: 'var(--primary)',      zIndex: 2 },
  // error:  { width: 1, color: 'var(--error)',        zIndex: 2 },
};
function rectStyle(rect) {
  return {
    position: "absolute",
    left: rect.x,
    top: rect.y,
    width: rect.width,
    height: rect.height
  };
}
function CellRingOverlay({ rect, kind, cellId }) {
  const ring = CELL_RING_STYLES[kind];
  const dataAttr = kind === "selected" && cellId ? { "data-selected-cell-id": cellId } : {};
  return /* @__PURE__ */ jsx(
    "div",
    {
      "aria-hidden": true,
      style: {
        ...rectStyle(rect),
        outline: `${ring.width}px solid ${ring.color}`,
        outlineOffset: `-${ring.width}px`,
        boxSizing: "border-box",
        pointerEvents: "none",
        transition: "none",
        zIndex: ring.zIndex
      },
      ...dataAttr
    }
  );
}
function ClipMask({ clipRect, children }) {
  return /* @__PURE__ */ jsx(
    "div",
    {
      "aria-hidden": true,
      style: {
        position: "absolute",
        left: clipRect.x,
        top: clipRect.y,
        width: clipRect.width,
        height: clipRect.height,
        overflow: "hidden",
        pointerEvents: "none"
      },
      children
    }
  );
}
function ActiveEditorHost({ rect, children }) {
  if (!children) {
    return /* @__PURE__ */ jsx(
      "div",
      {
        "aria-hidden": true,
        style: {
          ...rectStyle(rect),
          pointerEvents: "none",
          zIndex: 3,
          border: "1px dashed var(--primary)",
          boxSizing: "border-box"
        },
        "data-active-editor-host-scaffold": true
      }
    );
  }
  return /* @__PURE__ */ jsx(
    "div",
    {
      style: {
        ...rectStyle(rect),
        pointerEvents: "auto",
        zIndex: 3,
        boxSizing: "border-box",
        background: "var(--canvas)"
      },
      "data-active-editor-host": true,
      children
    }
  );
}
export {
  DataTableInteractionLayer
};
