// @qijenchen/design-system/dist/components/Tag/tag.js
import { jsxs, jsx } from "react/jsx-runtime";
import * as React from "react";
import { cva } from "class-variance-authority";
import { X } from "lucide-react";
import { cn } from "../../lib/utils.js";
import { Tooltip, TooltipTrigger, TooltipContent } from "../Tooltip/tooltip.js";
import { ItemInlineActionButton } from "../../patterns/element-anatomy/item-anatomy.js";
let _measureCtx = null;
function getMeasureCtx() {
  if (!_measureCtx) _measureCtx = document.createElement("canvas").getContext("2d");
  return _measureCtx;
}
const tagVariants = cva(
  "inline-flex items-center rounded-md border border-transparent transition-colors cursor-text",
  {
    variants: {
      color: {
        // 直接引用 primitive（bg=step-1, text=step-7），不經過語義層
        // step-1 在 dark mode 用 alpha 公式，step-7 用 lighter 公式——兩個 mode 都正確
        neutral: "bg-secondary text-foreground",
        blue: "bg-[var(--color-blue-1)] text-[var(--color-blue-7)]",
        red: "bg-[var(--color-deep-orange-1)] text-[var(--color-deep-orange-7)]",
        green: "bg-[var(--color-green-1)] text-[var(--color-green-7)]",
        yellow: "bg-[var(--color-yellow-1)] text-[var(--color-yellow-7)]",
        turquoise: "bg-[var(--color-turquoise-1)] text-[var(--color-turquoise-7)]",
        purple: "bg-[var(--color-purple-1)] text-[var(--color-purple-7)]",
        magenta: "bg-[var(--color-magenta-1)] text-[var(--color-magenta-7)]",
        indigo: "bg-[var(--color-indigo-1)] text-[var(--color-indigo-7)]"
      },
      size: {
        sm: "h-5 px-1 text-caption font-medium",
        md: "h-6 px-1 text-body font-normal",
        lg: "h-6 px-1 text-body font-normal"
      }
    },
    defaultVariants: {
      color: "neutral",
      size: "md"
    }
  }
);
const SOLID_CLASSES = {
  neutral: "bg-[var(--color-neutral-9)] text-inverse-fg",
  blue: "bg-[var(--color-blue-6)] text-on-emphasis",
  red: "bg-[var(--color-deep-orange-6)] text-on-emphasis",
  green: "bg-[var(--color-green-6)] text-on-emphasis",
  yellow: "bg-[var(--color-yellow-6)] text-[var(--warning-foreground)]",
  turquoise: "bg-[var(--color-turquoise-6)] text-on-emphasis",
  purple: "bg-[var(--color-purple-6)] text-on-emphasis",
  magenta: "bg-[var(--color-magenta-6)] text-on-emphasis",
  indigo: "bg-[var(--color-indigo-6)] text-on-emphasis"
};
const SOLID_DISMISS_HOVER = {
  neutral: { hover: "var(--inverse-neutral-hover)", active: "var(--inverse-neutral-active)" },
  blue: { hover: "var(--blue-hover)", active: "var(--blue-active)" },
  red: { hover: "var(--red-hover)", active: "var(--red-active)" },
  green: { hover: "var(--green-hover)", active: "var(--green-active)" },
  yellow: { hover: "var(--yellow-hover)", active: "var(--yellow-active)" },
  turquoise: { hover: "var(--turquoise-hover)", active: "var(--turquoise-active)" },
  purple: { hover: "var(--purple-hover)", active: "var(--purple-active)" },
  magenta: { hover: "var(--magenta-hover)", active: "var(--magenta-active)" },
  indigo: { hover: "var(--indigo-hover)", active: "var(--indigo-active)" }
};
function TagDismiss({ onDismiss, label, solid, color }) {
  const solidColors = solid && color ? SOLID_DISMISS_HOVER[color] : void 0;
  return /* @__PURE__ */ jsx(
    ItemInlineActionButton,
    {
      icon: X,
      size: "md",
      onClick: (e) => {
        e.stopPropagation();
        onDismiss();
      },
      "aria-label": `移除 ${label}`,
      style: solidColors ? { "--dismiss-hover": solidColors.hover, "--dismiss-active": solidColors.active } : void 0,
      hoverBgClassName: solidColors ? "group-hover/action:bg-[var(--dismiss-hover)] group-active/action:bg-[var(--dismiss-active)]" : void 0,
      className: "text-current hover:text-current active:text-current"
    }
  );
}
function TagInner({ className, color, size, icon: Icon, avatar, onDismiss, solid, unbounded = false, children, style, ...props }, forwardedRef) {
  const solidClass = solid ? SOLID_CLASSES[color ?? "neutral"] : void 0;
  const ownRef = React.useRef(null);
  const [isTruncated, setIsTruncated] = React.useState(false);
  React.useLayoutEffect(() => {
    const el = ownRef.current;
    if (!el) return;
    const ctx = getMeasureCtx();
    const check = () => {
      const textSpan = el.querySelector("[data-tag-text]");
      if (!textSpan || !ctx) return;
      const text = textSpan.textContent || "";
      const cs = getComputedStyle(textSpan);
      ctx.font = `${cs.fontWeight} ${cs.fontSize} ${cs.fontFamily}`;
      const textWidth = ctx.measureText(text).width;
      const padL = parseFloat(cs.paddingLeft) || 0;
      const padR = parseFloat(cs.paddingRight) || 0;
      const needed = textWidth + padL + padR;
      setIsTruncated(needed > textSpan.clientWidth + 1);
    };
    check();
    const obs = new ResizeObserver(check);
    obs.observe(el);
    return () => obs.disconnect();
  }, [children]);
  const label = typeof children === "string" ? children : "";
  const tag = /* @__PURE__ */ jsxs(
    "div",
    {
      ref: (el) => {
        ownRef.current = el;
        if (typeof forwardedRef === "function") forwardedRef(el);
        else if (forwardedRef) forwardedRef.current = el;
      },
      "data-tag-root": "",
      className: cn(tagVariants({ color, size }), solidClass, "w-fit min-w-0 overflow-hidden", className),
      style: {
        maxWidth: unbounded ? "var(--combobox-tag-area-inline-size, 100%)" : "min(var(--combobox-tag-area-inline-size, 10rem), 10rem)",
        ...style
      },
      ...props,
      children: [
        Icon && /* @__PURE__ */ jsx(Icon, { size: 16, "aria-hidden": true }),
        avatar && /* @__PURE__ */ jsx("span", { className: "shrink-0 w-4 h-4 rounded-full overflow-hidden inline-grid place-content-center [&>*]:w-full [&>*]:h-full", children: avatar }),
        /* @__PURE__ */ jsx("span", { "data-tag-text": "", className: "px-1 truncate min-w-0", children }),
        onDismiss && /* @__PURE__ */ jsx(TagDismiss, { onDismiss, label, solid, color: color ?? "neutral" })
      ]
    }
  );
  if (!isTruncated) return tag;
  return /* @__PURE__ */ jsxs(Tooltip, { children: [
    /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: tag }),
    /* @__PURE__ */ jsx(TooltipContent, { children })
  ] });
}
const Tag = React.forwardRef(TagInner);
Tag.displayName = "Tag";
const tagMeta = {
  component: "Tag",
  family: 3,
  variants: {
    neutral: { purpose: "通用分類、草稿、無特定語義" },
    blue: { purpose: "進行中、資訊提示、active 狀態（對應 --info）" },
    red: { purpose: "錯誤、已封鎖、危險（對應 --error）" },
    green: { purpose: "成功、已完成、已核准（對應 --success）" },
    yellow: { purpose: "警告、待審核、注意（對應 --warning）" },
    turquoise: { purpose: "分類色（無固定語義）" },
    purple: { purpose: "分類色（無固定語義）" },
    magenta: { purpose: "分類色（無固定語義）" },
    indigo: { purpose: "分類色（無固定語義）" }
  },
  sizes: {
    sm: { fieldHeight: 28, iconSize: 16, typography: "body" },
    md: { fieldHeight: 32, iconSize: 16, typography: "body" },
    lg: { fieldHeight: 40, iconSize: 20, typography: "body" }
  },
  states: ["default", "hover", "active", "focus-visible", "disabled"],
  tokens: {
    bg: ["bg-neutral-active", "bg-neutral-hover", "bg-secondary", "bg-transparent"],
    fg: ["text-foreground", "text-inverse-fg"],
    ring: []
  },
  defaultVariant: "neutral",
  defaultSize: "md"
};
export {
  Tag,
  tagMeta,
  tagVariants
};
