// @qijenchen/design-system/dist/components/Sheet/sheet.js
import { jsx, jsxs } from "react/jsx-runtime";
import * as React from "react";
import * as DialogPrimitive from "@radix-ui/react-dialog";
import { cva } from "class-variance-authority";
import { X } from "lucide-react";
import { cn } from "../../lib/utils.js";
import { SurfaceFooter, SurfaceHeader } from "../../patterns/overlay-surface/overlay-surface.js";
import { Button } from "../Button/button.js";
import { ScrollArea } from "../ScrollArea/scroll-area.js";
const Sheet = DialogPrimitive.Root;
const SheetTrigger = DialogPrimitive.Trigger;
const SheetClose = DialogPrimitive.Close;
const SheetPortal = DialogPrimitive.Portal;
const SheetOverlay = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Overlay,
  {
    className: cn(
      "fixed inset-0 z-50 bg-overlay data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0",
      className
    ),
    ...props,
    ref
  }
));
SheetOverlay.displayName = DialogPrimitive.Overlay.displayName;
const sheetVariants = cva(
  // 核心容器 — 無 padding(由 SheetBody / SheetHeader / SheetFooter 自理 padding,
  // 對齊 overlay-surface pattern + Dialog canonical)
  // Animation canonical:300ms 雙向一致(D4 audit:500ms 太久 sluggish)+ motion-reduce 豁免
  "fixed z-50 flex flex-col bg-surface-raised shadow-[var(--elevation-200)] transition ease-in-out data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:duration-300 data-[state=open]:duration-300 motion-reduce:transition-none motion-reduce:data-[state=open]:duration-0 motion-reduce:data-[state=closed]:duration-0",
  {
    variants: {
      side: {
        top: "inset-x-0 top-0 border-b border-divider data-[state=closed]:slide-out-to-top data-[state=open]:slide-in-from-top",
        bottom: "inset-x-0 bottom-0 border-t border-divider data-[state=closed]:slide-out-to-bottom data-[state=open]:slide-in-from-bottom",
        left: "inset-y-0 left-0 h-full w-3/4 border-r border-divider data-[state=closed]:slide-out-to-left data-[state=open]:slide-in-from-left sm:max-w-md",
        right: "inset-y-0 right-0 h-full w-3/4 border-l border-divider data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-md"
      }
    },
    defaultVariants: {
      side: "right"
    }
  }
);
const handleSheetOpenAutoFocus = (e) => {
  e.preventDefault();
  const content = e.currentTarget;
  const firstBodyTarget = content.querySelector(
    "[data-sheet-body] input:not([disabled]),[data-sheet-body] textarea:not([disabled]),[data-sheet-body] select:not([disabled]),[data-sheet-body] button:not([disabled]):not([data-dismiss])"
  );
  const firstFooterButton = content.querySelector(
    "[data-sheet-footer] button:not([disabled]):not([data-dismiss])"
  );
  (firstBodyTarget ?? firstFooterButton ?? content).focus({ preventScroll: true });
};
const SheetContent = React.forwardRef(({ side = "right", className, children, ...props }, ref) => /* @__PURE__ */ jsxs(SheetPortal, { children: [
  /* @__PURE__ */ jsx(SheetOverlay, {}),
  /* @__PURE__ */ jsx(
    DialogPrimitive.Content,
    {
      ref,
      onOpenAutoFocus: handleSheetOpenAutoFocus,
      className: cn(sheetVariants({ side }), className),
      ...props,
      children
    }
  )
] }));
SheetContent.displayName = DialogPrimitive.Content.displayName;
const SheetHeader = React.forwardRef(({ className, children, ...props }, ref) => (
  // 2026-05-18:className 不再硬加 justify-between(同 DialogHeader 邏輯,避 column mode 破裂)。
  /* @__PURE__ */ jsxs(
    SurfaceHeader,
    {
      ref,
      className,
      ...props,
      children: [
        /* @__PURE__ */ jsx("div", { className: "flex-1 min-w-0", children }),
        /* @__PURE__ */ jsx(DialogPrimitive.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { "data-dismiss": true, iconOnly: true, dismiss: true, size: "sm", startIcon: X, "aria-label": "關閉" }) })
      ]
    }
  )
));
SheetHeader.displayName = "SheetHeader";
const SheetBody = React.forwardRef(({ className, children, ...props }, ref) => /* @__PURE__ */ jsx(ScrollArea, { ref, "data-sheet-body": true, className: "flex-1 min-h-0", ...props, children: /* @__PURE__ */ jsx(
  "div",
  {
    className: cn(
      "px-[var(--layout-space-loose)] pt-[var(--layout-space-tight)] pb-[var(--layout-space-bottom)]",
      className
    ),
    children
  }
) }));
SheetBody.displayName = "SheetBody";
const SheetFooter = React.forwardRef(({ ...props }, ref) => /* @__PURE__ */ jsx(SurfaceFooter, { ref, "data-sheet-footer": true, ...props }));
SheetFooter.displayName = "SheetFooter";
const SheetTitle = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Title,
  {
    ref,
    className: cn("text-body-lg font-medium truncate text-foreground", className),
    ...props
  }
));
SheetTitle.displayName = DialogPrimitive.Title.displayName;
const SheetDescription = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Description,
  {
    ref,
    className: cn("mt-[var(--item-gap-label-desc-reading-lg)] text-body text-fg-secondary", className),
    ...props
  }
));
SheetDescription.displayName = DialogPrimitive.Description.displayName;
const sheetMeta = {
  component: "Sheet",
  family: null,
  // non-family composite / overlay / layout
  variants: {},
  sizes: {},
  states: ["default", "hover", "active", "focus-visible", "disabled"],
  tokens: {
    bg: ["bg-surface-raised"],
    fg: ["text-fg-secondary", "text-foreground"],
    ring: []
  }
};
export {
  Sheet,
  SheetBody,
  SheetClose,
  SheetContent,
  SheetDescription,
  SheetFooter,
  SheetHeader,
  SheetOverlay,
  SheetPortal,
  SheetTitle,
  SheetTrigger,
  sheetMeta,
  sheetVariants
};
