// @qijenchen/design-system/dist/patterns/element-anatomy/item-anatomy.js
import { jsx, jsxs } from "react/jsx-runtime";
import * as React from "react";
import { cva } from "class-variance-authority";
import { cn } from "../../lib/utils.js";
import { Avatar } from "../../components/Avatar/avatar.js";
import { Tooltip, TooltipTrigger, TooltipContent } from "../../components/Tooltip/tooltip.js";
const ICON_SIZE = {
  sm: 16,
  md: 16,
  lg: 20
};
const AVATAR_SIZE = {
  inline: { sm: 20, md: 24, lg: 24 },
  block: { sm: 32, md: 32, lg: 40 }
};
function getUniformPrefixSlotStyle(size) {
  return { "--item-prefix-slot": `${AVATAR_SIZE.inline[size]}px` };
}
const INLINE_ACTION_HOVER_BG_SIZE = {
  sm: 18,
  md: 18,
  lg: 22
};
const ROW_PADDING_BY_SIZE = {
  sm: "text-body leading-compact py-[calc((var(--field-height-sm)-1lh)/2)]",
  md: "text-body leading-compact py-[calc((var(--field-height-md)-1lh)/2)]",
  lg: "text-body-lg leading-compact py-[calc((var(--field-height-lg)-1lh)/2)]"
};
const ItemPrefix = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  "span",
  {
    ref,
    className: cn(
      "h-[1lh] shrink-0 flex items-center",
      // Uniform prefix slot:讀取 `--item-prefix-slot` CSS variable,
      // fallback `auto`(= 預設行為,不佔額外寬度、不 center)。
      // Row-primitive 頂層容器(例如 `<SidebarProvider uniformPrefix>` 的 wrapper)
      // 若設定這個 var,小尺寸 prefix(icon 16px)會在固定寬槽(24px @ md)內置中,
      // 與大尺寸 prefix(app logo 24px)對齊,達成 label 齊左。
      // 詳見 item-anatomy.spec.md「Uniform prefix slot」節。
      "min-w-[var(--item-prefix-slot,auto)] justify-center",
      className
    ),
    ...props
  }
));
ItemPrefix.displayName = "ItemPrefix";
const itemPrefixAlignVariants = cva(
  "flex items-center gap-2 shrink-0",
  {
    variants: {
      align: {
        inline: "h-[1lh]",
        // sm/md desc = text-caption (12px × 1.3 = 15.6px)
        // block-sm / block-md:scanning mode(body + caption 組合)→ 用 scanning mode 專屬 gap token
        "block-sm": "h-[calc(1lh+var(--item-gap-label-desc-scanning)+var(--font-caption-size)*1.3)]",
        "block-md": "h-[calc(1lh+var(--item-gap-label-desc-scanning)+var(--font-caption-size)*1.3)]",
        // block-lg:scanning-lg mode(body-lg + body compact 組合)→ 用 scanning-lg mode 專屬 gap token
        "block-lg": "h-[calc(1lh+var(--item-gap-label-desc-scanning-lg)+var(--font-body-size)*1.3)]"
      }
    },
    defaultVariants: {
      align: "inline"
    }
  }
);
const ItemLabel = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  "span",
  {
    ref,
    "data-sidebar": "menu-label",
    className: cn("min-w-0 flex-1 truncate", className),
    ...props
  }
));
ItemLabel.displayName = "ItemLabel";
const ItemContent = React.forwardRef(
  ({
    label,
    description,
    mode = "reading",
    size = "md",
    descriptionTone = "secondary",
    descriptionWrap = true,
    descriptionClamp,
    descriptionBreakWords = false,
    labelTruncate = true,
    labelClassName,
    descriptionClassName,
    className,
    ...props
  }, ref) => {
    const toneClass = {
      secondary: "text-fg-secondary",
      error: "text-error-text",
      muted: "text-fg-muted",
      disabled: "text-fg-disabled"
    }[descriptionTone];
    const isLg = size === "lg";
    const modeClass = mode === "scanning" ? isLg ? "text-[length:var(--font-body-size)] leading-compact" : "text-[length:var(--font-caption-size)] leading-compact" : isLg ? "text-[length:var(--font-body-size)]" : "";
    const clampClass = descriptionClamp ? `line-clamp-${descriptionClamp}` : "";
    return /* @__PURE__ */ jsxs(
      "div",
      {
        ref,
        className: cn("flex flex-col min-w-0 flex-1", className),
        ...props,
        children: [
          /* @__PURE__ */ jsx("span", { className: cn(labelTruncate && "truncate", labelClassName), children: label }),
          description && /* @__PURE__ */ jsx(
            "span",
            {
              className: cn(
                // Typography-mode-aware gap token(2026-04-23):
                // (mode, size)2 維正交 → 4 token 對應 4 typography 組合
                mode === "scanning" ? isLg ? "mt-[var(--item-gap-label-desc-scanning-lg)]" : "mt-[var(--item-gap-label-desc-scanning)]" : isLg ? "mt-[var(--item-gap-label-desc-reading-lg)]" : "mt-[var(--item-gap-label-desc-reading)]",
                modeClass,
                toneClass,
                clampClass,
                descriptionBreakWords && "break-words",
                !descriptionWrap && "truncate",
                descriptionClassName
              ),
              children: description
            }
          )
        ]
      }
    );
  }
);
ItemContent.displayName = "ItemContent";
const RowSizeContext = React.createContext(null);
const RowSizeProvider = RowSizeContext.Provider;
function useRowSize(fallback = "md") {
  return React.useContext(RowSizeContext) ?? fallback;
}
const ItemIcon = React.forwardRef(
  ({ icon: Icon, className, ...props }, ref) => {
    const size = useRowSize();
    return (
      // data-prefix-type="icon" 讓 row-primitive 容器的 `:has()` selector 能偵測類型,
      // 用於 `<SidebarProvider uniformPrefix>` opt-in 後的混用 auto-detection
      /* @__PURE__ */ jsx(ItemPrefix, { "data-prefix-type": "icon", children: /* @__PURE__ */ jsx(
        Icon,
        {
          ref,
          size: ICON_SIZE[size],
          className: cn("shrink-0", className),
          "aria-hidden": true,
          ...props
        }
      ) })
    );
  }
);
ItemIcon.displayName = "ItemIcon";
const ItemAvatar = React.forwardRef(
  ({ mode = "inline", ...props }, ref) => {
    const size = useRowSize();
    return (
      // data-prefix-type="avatar" 讓 row-primitive 容器的 `:has()` selector 能偵測類型,
      // 用於 `<SidebarProvider uniformPrefix>` opt-in 後的混用 auto-detection
      /* @__PURE__ */ jsx(ItemPrefix, { "data-prefix-type": "avatar", children: /* @__PURE__ */ jsx(Avatar, { ref, size: AVATAR_SIZE[mode][size], ...props }) })
    );
  }
);
ItemAvatar.displayName = "ItemAvatar";
const ItemInlineActionButton = React.forwardRef(({ icon: Icon, className, iconClassName, style, type = "button", size: sizeProp, hoverBgClassName, overlayTrigger = false, ...rest }, ref) => {
  const contextSize = useRowSize();
  const size = sizeProp ?? contextSize;
  const iconPx = ICON_SIZE[size];
  const hoverBgPx = INLINE_ACTION_HOVER_BG_SIZE[size];
  return /* @__PURE__ */ jsxs(
    "button",
    {
      ref,
      type,
      className: cn(
        "group/action relative grid place-content-center shrink-0 cursor-pointer",
        "text-fg-muted hover:text-foreground active:text-foreground transition-colors",
        // Overlay trigger active state — 只在 consumer 顯式宣告 overlayTrigger=true 時生效
        // (canonical 2026-05-05:Collapsible / drag / dismiss 等 in-place 互動 ≠ overlay,
        // 不應 leak 視覺 lock。詳 inline-action.spec.md「Overlay trigger canonical」)
        overlayTrigger && "data-[state=open]:text-foreground",
        "focus-visible:outline-2 focus-visible:outline-ring",
        className
      ),
      style: { width: iconPx, height: iconPx, ...style },
      ...rest,
      children: [
        /* @__PURE__ */ jsx(
          "span",
          {
            "aria-hidden": true,
            className: cn(
              "absolute pointer-events-none",
              "rounded-md",
              "bg-transparent",
              hoverBgClassName ?? "group-hover/action:bg-neutral-hover group-active/action:bg-neutral-active",
              // Overlay 開啟 = 維持 host hover bg(只在 overlayTrigger=true 時生效)
              overlayTrigger && "group-data-[state=open]/action:bg-neutral-hover",
              "transition-colors"
            ),
            style: {
              width: hoverBgPx,
              height: hoverBgPx,
              top: "50%",
              left: "50%",
              transform: "translate(-50%, -50%)"
            }
          }
        ),
        /* @__PURE__ */ jsx(Icon, { size: iconPx, className: cn("relative", iconClassName), "aria-hidden": true })
      ]
    }
  );
});
ItemInlineActionButton.displayName = "ItemInlineActionButton";
const ItemInlineAction = React.forwardRef(({ action, size }, ref) => /* @__PURE__ */ jsxs(Tooltip, { children: [
  /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: /* @__PURE__ */ jsx(
    ItemInlineActionButton,
    {
      ref,
      icon: action.icon,
      "aria-label": action.label,
      onClick: action.onClick,
      size
    }
  ) }),
  /* @__PURE__ */ jsx(TooltipContent, { children: action.label })
] }));
ItemInlineAction.displayName = "ItemInlineAction";
const SUFFIX_HOVER_REVEAL_BY_GROUP = {
  "menu-item": "opacity-0 group-hover/menu-item:opacity-100 group-has-[:focus-visible]/menu-item:opacity-100 transition-opacity duration-150",
  "tree-item": "opacity-0 group-hover/tree-item:opacity-100 group-has-[:focus-visible]/tree-item:opacity-100 transition-opacity duration-150",
  row: "opacity-0 group-hover/row:opacity-100 group-has-[:focus-visible]/row:opacity-100 transition-opacity duration-150"
};
const ItemSuffix = React.forwardRef(
  ({ className, hoverReveal = false, hoverGroup = "menu-item", ...props }, ref) => /* @__PURE__ */ jsx(
    "span",
    {
      ref,
      className: cn(
        "h-[1lh] shrink-0 ml-auto flex items-center gap-2",
        hoverReveal && SUFFIX_HOVER_REVEAL_BY_GROUP[hoverGroup],
        className
      ),
      ...props
    }
  )
);
ItemSuffix.displayName = "ItemSuffix";
export {
  AVATAR_SIZE,
  ICON_SIZE,
  INLINE_ACTION_HOVER_BG_SIZE,
  ItemAvatar,
  ItemContent,
  ItemIcon,
  ItemInlineAction,
  ItemInlineActionButton,
  ItemLabel,
  ItemPrefix,
  ItemSuffix,
  ROW_PADDING_BY_SIZE,
  RowSizeProvider,
  getUniformPrefixSlotStyle,
  itemPrefixAlignVariants,
  useRowSize
};
