// @qijenchen/design-system/dist/index.js
import { Accordion, AccordionContent, AccordionItem, AccordionTrigger, accordionMeta } from "./components/Accordion/accordion.js";
import { Alert, alertMeta, alertVariants } from "./components/Alert/alert.js";
import { AppShell, AppShellAside, useAppShell } from "./components/AppShell/app-shell.js";
import { AspectRatio, aspectRatioMeta } from "./components/AspectRatio/aspect-ratio.js";
import { Avatar, avatarMeta } from "./components/Avatar/avatar.js";
import { Badge, badgeMeta, badgeVariants } from "./components/Badge/badge.js";
import { Breadcrumb, BreadcrumbEllipsis, BreadcrumbItem, BreadcrumbLink, BreadcrumbList, BreadcrumbPage, BreadcrumbSeparator, breadcrumbMeta } from "./components/Breadcrumb/breadcrumb.js";
import { BULK_ACTION_BAR_DEFAULT_LABELS, BulkActionBar, bulkActionBarMeta } from "./components/BulkActionBar/bulk-action-bar.js";
import { Button, ButtonGroupContext, buttonMeta, buttonVariants } from "./components/Button/button.js";
import { Calendar, calendarMeta } from "./components/Calendar/calendar.js";
import { Carousel, CarouselContent, CarouselDots, CarouselItem, CarouselNext, CarouselPrevious, carouselMeta } from "./components/Carousel/carousel.js";
import { ChartContainer, ChartLegend, ChartLegendContent, ChartStyle, ChartTooltip, ChartTooltipContent, chartMeta } from "./components/Chart/chart.js";
import { Checkbox, checkboxMeta, checkboxVariants } from "./components/Checkbox/checkbox.js";
import { Chip, ChipGroup, chipMeta, chipVariants } from "./components/Chip/chip.js";
import { CircularProgress, circularProgressMeta } from "./components/CircularProgress/circular-progress.js";
import { Coachmark, coachmarkMeta } from "./components/Coachmark/coachmark.js";
import { Combobox, comboboxMeta } from "./components/Combobox/combobox.js";
import { Command, CommandDialog, CommandEmpty, CommandGroup, CommandInput, CommandItem, CommandList, CommandSeparator, CommandShortcut, commandMeta } from "./components/Command/command.js";
import { DataTable, MIN_COLUMN_WIDTH, dataTableMeta, dataTableVariants } from "./components/DataTable/data-table.js";
import { DateGrid, dateGridMeta } from "./components/DateGrid/date-grid.js";
import { DatePicker, DatePickerRange, datePickerMeta, formatDate } from "./components/DatePicker/date-picker.js";
import { DescriptionItem, DescriptionList, descriptionListMeta } from "./components/DescriptionList/description-list.js";
import { Dialog, DialogBody, DialogClose, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogOverlay, DialogPortal, DialogTitle, DialogTrigger, dialogMeta } from "./components/Dialog/dialog.js";
import { DropdownMenu, DropdownMenuCheckboxItem, DropdownMenuContent, DropdownMenuGroup, DropdownMenuItem, DropdownMenuLabel, DropdownMenuPortal, DropdownMenuRadioGroup, DropdownMenuRadioItem, DropdownMenuSeparator, DropdownMenuShortcut, DropdownMenuSub, DropdownMenuSubContent, DropdownMenuSubTrigger, DropdownMenuTrigger, dropdownMenuMeta, floatingLayerClass } from "./components/DropdownMenu/dropdown-menu.js";
import { Empty, emptyMeta } from "./components/Empty/empty.js";
import { Field, FieldDescription, FieldError, FieldGroup, FieldLabel, fieldMeta } from "./components/Field/field.js";
import { FieldControlGroup, fieldControlGroupMeta } from "./components/FieldControlGroup/field-control-group.js";
import { FileItem, fileItemMeta } from "./components/FileItem/file-item.js";
import { FileUpload, fileUploadMeta } from "./components/FileUpload/file-upload.js";
import { FileViewer, fileViewerMeta, registerFileRenderer } from "./components/FileViewer/file-viewer.js";
import { HoverCard, HoverCardContent, HoverCardTrigger, hoverCardMeta } from "./components/HoverCard/hover-card.js";
import { Input, inputMeta } from "./components/Input/input.js";
import { LinkInput, linkInputMeta } from "./components/LinkInput/link-input.js";
import { MenuFooter, MenuGroup, MenuItem, menuItemMeta, menuItemVariants } from "./components/Menu/menu-item.js";
import { NAMECARD_DEFAULT_FIELD_KEYS, NameCard, NameCardDefaultActions, nameCardMeta } from "./components/NameCard/name-card.js";
import { Notice, SUBTLE_ICON_COLOR, noticeMeta, useInverseTheme } from "./components/Notice/notice.js";
import { NumberInput, formatNumber, numberInputMeta } from "./components/NumberInput/number-input.js";
import { OverflowIndicator, overflowIndicatorMeta } from "./components/OverflowIndicator/overflow-indicator.js";
import { PeoplePicker, peoplePickerMeta } from "./components/PeoplePicker/people-picker.js";
import { Popover, PopoverAnchor, PopoverBody, PopoverClose, PopoverContent, PopoverFooter, PopoverHeader, PopoverTitle, PopoverTrigger, popoverMeta } from "./components/Popover/popover.js";
import { ProgressBar, progressBarMeta } from "./components/ProgressBar/progress-bar.js";
import { RadioGroup, RadioGroupItem, radioGroupMeta, radioItemVariants } from "./components/RadioGroup/radio-group.js";
import { Rating, ratingMeta } from "./components/Rating/rating.js";
import { ScrollArea, ScrollBar, scrollAreaMeta } from "./components/ScrollArea/scroll-area.js";
import { SegmentedControl, SegmentedControlItem, segmentedControlItemVariants, segmentedControlMeta, segmentedControlVariants } from "./components/SegmentedControl/segmented-control.js";
import { Select, selectMeta } from "./components/Select/select.js";
import { SelectMenu, selectMenuMeta } from "./components/SelectMenu/select-menu.js";
import { SelectionItem, selectionItemMeta, selectionItemStyles } from "./components/SelectionControl/selection-item.js";
import { Separator, separatorMeta } from "./components/Separator/separator.js";
import { Sheet, SheetBody, SheetClose, SheetContent, SheetDescription, SheetFooter, SheetHeader, SheetOverlay, SheetPortal, SheetTitle, SheetTrigger, sheetMeta, sheetVariants } from "./components/Sheet/sheet.js";
import { Sidebar, SidebarContent, SidebarFooter, SidebarGroup, SidebarGroupAction, SidebarGroupContent, SidebarGroupLabel, SidebarHeader, SidebarInput, SidebarMenu, SidebarMenuAction, SidebarMenuBadge, SidebarMenuButton, SidebarMenuItem, SidebarMenuSkeleton, SidebarProvider, SidebarSeparator, SidebarTrigger, sidebarMeta, useSidebar } from "./components/Sidebar/sidebar.js";
import { Skeleton, skeletonMeta } from "./components/Skeleton/skeleton.js";
import { Slider, sliderMeta, sliderRootVariants } from "./components/Slider/slider.js";
import { StepContent, StepDescription, StepItem, StepLabel, Steps, stepItemVariants, stepsMeta, stepsRootVariants } from "./components/Steps/steps.js";
import { Switch, switchMeta, switchVariants } from "./components/Switch/switch.js";
import { Tabs, TabsContent, TabsList, TabsTrigger, tabsMeta, tabsTriggerVariants } from "./components/Tabs/tabs.js";
import { Tag, tagMeta, tagVariants } from "./components/Tag/tag.js";
import { Textarea, textareaMeta, textareaVariants } from "./components/Textarea/textarea.js";
import { TimePicker, timePickerMeta } from "./components/TimePicker/time-picker.js";
import { Toaster, toast, toastMeta } from "./components/Toast/toast.js";
import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger, tooltipMeta } from "./components/Tooltip/tooltip.js";
import { TreeItem, TreeView, treeItemVariants, treeViewMeta } from "./components/TreeView/tree-view.js";
import { AVATAR_SIZE, ICON_SIZE, INLINE_ACTION_HOVER_BG_SIZE, ItemAvatar, ItemContent, ItemIcon, ItemInlineAction, ItemInlineActionButton, ItemLabel, ItemPrefix, ItemSuffix, ROW_PADDING_BY_SIZE, RowSizeProvider, getUniformPrefixSlotStyle, itemPrefixAlignVariants, useRowSize } from "./patterns/element-anatomy/item-anatomy.js";
import { ARROW_BUTTON_WIDTH, FADE_WIDTH, OverflowMenuTriggerButton, OverflowScrollArrow, SCROLL_PAGE_RATIO, buildFadeMask, useScrollByPage } from "./patterns/horizontal-overflow/horizontal-overflow.js";
import { SurfaceBody, SurfaceFooter, SurfaceHeader } from "./patterns/overlay-surface/overlay-surface.js";
import { ResizeHandle } from "./patterns/resize-handle/resize-handle.js";
import { useControllable } from "./hooks/use-controllable.js";
import { useIsNarrowViewport } from "./hooks/use-is-narrow-viewport.js";
import { useIsTouchDevice } from "./hooks/use-is-touch-device.js";
import { useOverflowIndices, useScrollEdges } from "./hooks/use-overflow-items.js";
import { dragActiveCursor, dragHandleCursorClass, dragSourceClass, dragSourceStyle, dropIndicatorColumn, dropIndicatorInside, dropIndicatorRow, isReorderNoop, reconstructFullRowGhost, snapToCursorModifier } from "./lib/drag-visual.js";
import { applySelectAll, clearSelection } from "./lib/multi-select-ordering.js";
import { cn } from "./lib/utils.js";
import { PEOPLE_PICKER_LENGTH1_WRAPPER_CLASS, getPeoplePickerTagWrapperClass } from "./components/PeoplePicker/people-picker-helpers.js";
export {
  ARROW_BUTTON_WIDTH,
  AVATAR_SIZE,
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionTrigger,
  Alert,
  AppShell,
  AppShellAside,
  AspectRatio,
  Avatar,
  BULK_ACTION_BAR_DEFAULT_LABELS,
  Badge,
  Breadcrumb,
  BreadcrumbEllipsis,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
  BulkActionBar,
  Button,
  ButtonGroupContext,
  Calendar,
  Carousel,
  CarouselContent,
  CarouselDots,
  CarouselItem,
  CarouselNext,
  CarouselPrevious,
  ChartContainer,
  ChartLegend,
  ChartLegendContent,
  ChartStyle,
  ChartTooltip,
  ChartTooltipContent,
  Checkbox,
  Chip,
  ChipGroup,
  CircularProgress,
  Coachmark,
  Combobox,
  Command,
  CommandDialog,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator,
  CommandShortcut,
  DataTable,
  DateGrid,
  DatePicker,
  DatePickerRange,
  DescriptionItem,
  DescriptionList,
  Dialog,
  DialogBody,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogOverlay,
  DialogPortal,
  DialogTitle,
  DialogTrigger,
  DropdownMenu,
  DropdownMenuCheckboxItem,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuPortal,
  DropdownMenuRadioGroup,
  DropdownMenuRadioItem,
  DropdownMenuSeparator,
  DropdownMenuShortcut,
  DropdownMenuSub,
  DropdownMenuSubContent,
  DropdownMenuSubTrigger,
  DropdownMenuTrigger,
  Empty,
  FADE_WIDTH,
  Field,
  FieldControlGroup,
  FieldDescription,
  FieldError,
  FieldGroup,
  FieldLabel,
  FileItem,
  FileUpload,
  FileViewer,
  HoverCard,
  HoverCardContent,
  HoverCardTrigger,
  ICON_SIZE,
  INLINE_ACTION_HOVER_BG_SIZE,
  Input,
  ItemAvatar,
  ItemContent,
  ItemIcon,
  ItemInlineAction,
  ItemInlineActionButton,
  ItemLabel,
  ItemPrefix,
  ItemSuffix,
  LinkInput,
  MIN_COLUMN_WIDTH,
  MenuFooter,
  MenuGroup,
  MenuItem,
  NAMECARD_DEFAULT_FIELD_KEYS,
  NameCard,
  NameCardDefaultActions,
  Notice,
  NumberInput,
  OverflowIndicator,
  OverflowMenuTriggerButton,
  OverflowScrollArrow,
  PEOPLE_PICKER_LENGTH1_WRAPPER_CLASS,
  PeoplePicker,
  Popover,
  PopoverAnchor,
  PopoverBody,
  PopoverClose,
  PopoverContent,
  PopoverFooter,
  PopoverHeader,
  PopoverTitle,
  PopoverTrigger,
  ProgressBar,
  ROW_PADDING_BY_SIZE,
  RadioGroup,
  RadioGroupItem,
  Rating,
  ResizeHandle,
  RowSizeProvider,
  SCROLL_PAGE_RATIO,
  SUBTLE_ICON_COLOR,
  ScrollArea,
  ScrollBar,
  SegmentedControl,
  SegmentedControlItem,
  Select,
  SelectMenu,
  SelectionItem,
  Separator,
  Sheet,
  SheetBody,
  SheetClose,
  SheetContent,
  SheetDescription,
  SheetFooter,
  SheetHeader,
  SheetOverlay,
  SheetPortal,
  SheetTitle,
  SheetTrigger,
  Sidebar,
  SidebarContent,
  SidebarFooter,
  SidebarGroup,
  SidebarGroupAction,
  SidebarGroupContent,
  SidebarGroupLabel,
  SidebarHeader,
  SidebarInput,
  SidebarMenu,
  SidebarMenuAction,
  SidebarMenuBadge,
  SidebarMenuButton,
  SidebarMenuItem,
  SidebarMenuSkeleton,
  SidebarProvider,
  SidebarSeparator,
  SidebarTrigger,
  Skeleton,
  Slider,
  StepContent,
  StepDescription,
  StepItem,
  StepLabel,
  Steps,
  SurfaceBody,
  SurfaceFooter,
  SurfaceHeader,
  Switch,
  Tabs,
  TabsContent,
  TabsList,
  TabsTrigger,
  Tag,
  Textarea,
  TimePicker,
  Toaster,
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
  TreeItem,
  TreeView,
  accordionMeta,
  alertMeta,
  alertVariants,
  applySelectAll,
  aspectRatioMeta,
  avatarMeta,
  badgeMeta,
  badgeVariants,
  breadcrumbMeta,
  buildFadeMask,
  bulkActionBarMeta,
  buttonMeta,
  buttonVariants,
  calendarMeta,
  carouselMeta,
  chartMeta,
  checkboxMeta,
  checkboxVariants,
  chipMeta,
  chipVariants,
  circularProgressMeta,
  clearSelection,
  cn,
  coachmarkMeta,
  comboboxMeta,
  commandMeta,
  dataTableMeta,
  dataTableVariants,
  dateGridMeta,
  datePickerMeta,
  descriptionListMeta,
  dialogMeta,
  dragActiveCursor,
  dragHandleCursorClass,
  dragSourceClass,
  dragSourceStyle,
  dropIndicatorColumn,
  dropIndicatorInside,
  dropIndicatorRow,
  dropdownMenuMeta,
  emptyMeta,
  fieldControlGroupMeta,
  fieldMeta,
  fileItemMeta,
  fileUploadMeta,
  fileViewerMeta,
  floatingLayerClass,
  formatDate,
  formatNumber,
  getPeoplePickerTagWrapperClass,
  getUniformPrefixSlotStyle,
  hoverCardMeta,
  inputMeta,
  isReorderNoop,
  itemPrefixAlignVariants,
  linkInputMeta,
  menuItemMeta,
  menuItemVariants,
  nameCardMeta,
  noticeMeta,
  numberInputMeta,
  overflowIndicatorMeta,
  peoplePickerMeta,
  popoverMeta,
  progressBarMeta,
  radioGroupMeta,
  radioItemVariants,
  ratingMeta,
  reconstructFullRowGhost,
  registerFileRenderer,
  scrollAreaMeta,
  segmentedControlItemVariants,
  segmentedControlMeta,
  segmentedControlVariants,
  selectMenuMeta,
  selectMeta,
  selectionItemMeta,
  selectionItemStyles,
  separatorMeta,
  sheetMeta,
  sheetVariants,
  sidebarMeta,
  skeletonMeta,
  sliderMeta,
  sliderRootVariants,
  snapToCursorModifier,
  stepItemVariants,
  stepsMeta,
  stepsRootVariants,
  switchMeta,
  switchVariants,
  tabsMeta,
  tabsTriggerVariants,
  tagMeta,
  tagVariants,
  textareaMeta,
  textareaVariants,
  timePickerMeta,
  toast,
  toastMeta,
  tooltipMeta,
  treeItemVariants,
  treeViewMeta,
  useAppShell,
  useControllable,
  useInverseTheme,
  useIsNarrowViewport,
  useIsTouchDevice,
  useOverflowIndices,
  useRowSize,
  useScrollByPage,
  useScrollEdges,
  useSidebar
};
