// @qijenchen/design-system/dist/components/Field/field-context.js
import * as React from "react";
const FieldContext = React.createContext(null);
function useFieldContext() {
  return React.useContext(FieldContext);
}
const FieldSurfaceContext = React.createContext(null);
function useFieldSurface() {
  const surfaceCtx = React.useContext(FieldSurfaceContext);
  if (surfaceCtx) return surfaceCtx;
  const fieldCtx = React.useContext(FieldContext);
  return (fieldCtx == null ? void 0 : fieldCtx.surface) ?? "form";
}
function FieldSurfaceProvider({
  surface,
  children
}) {
  return React.createElement(FieldSurfaceContext.Provider, { value: surface }, children);
}
const TableScrollContext = React.createContext(false);
function useTableIsScrolling() {
  return React.useContext(TableScrollContext);
}
function TableScrollProvider({
  isScrolling,
  children
}) {
  return React.createElement(TableScrollContext.Provider, { value: isScrolling }, children);
}
export {
  FieldContext,
  FieldSurfaceProvider,
  TableScrollProvider,
  useFieldContext,
  useFieldSurface,
  useTableIsScrolling
};
