// @qijenchen/design-system/dist/components/FileViewer/image-renderer.js
import { jsx } from "react/jsx-runtime";
import * as React from "react";
import { TransformWrapper, TransformComponent } from "react-zoom-pan-pinch";
const MIN_SCALE = 0.1;
const MAX_SCALE = 4;
const ImageRenderer = ({
  file,
  zoom,
  onZoomChange,
  fitRequest,
  onCapabilitiesChange
}) => {
  const apiRef = React.useRef(null);
  const imgRef = React.useRef(null);
  const containerRef = React.useRef(null);
  const [loaded, setLoaded] = React.useState(false);
  const handleImageLoadRafIdRef = React.useRef(0);
  React.useEffect(() => () => {
    if (handleImageLoadRafIdRef.current) cancelAnimationFrame(handleImageLoadRafIdRef.current);
  }, []);
  const lastFitModeRef = React.useRef("fit-page");
  const programmaticZoomRef = React.useRef(false);
  React.useEffect(() => {
    onCapabilitiesChange({ zoom: true });
  }, [onCapabilitiesChange]);
  React.useEffect(() => {
    setLoaded(false);
    lastFitModeRef.current = "fit-page";
    const img = imgRef.current;
    if (img && img.complete && img.naturalWidth > 0) {
      Promise.resolve().then(() => handleImageLoad());
    }
  }, [file.url]);
  const computeFitScale = React.useCallback((fit) => {
    const img = imgRef.current;
    const container = containerRef.current;
    if (!img || !container) return null;
    if (!img.naturalWidth || !img.naturalHeight) return null;
    const cw = container.clientWidth;
    const ch = container.clientHeight;
    if (cw <= 0 || ch <= 0) return null;
    const widthRatio = cw / img.naturalWidth;
    const heightRatio = ch / img.naturalHeight;
    return fit === "fit-width" ? widthRatio : Math.min(widthRatio, heightRatio);
  }, []);
  const clampToPct = React.useCallback((scale) => {
    const clamped = Math.min(MAX_SCALE, Math.max(MIN_SCALE, scale));
    return Math.floor(clamped * 100);
  }, []);
  const handleImageLoad = React.useCallback(() => {
    setLoaded(true);
    const scale = computeFitScale("fit-page");
    if (scale == null) return;
    const pct = clampToPct(scale);
    lastFitModeRef.current = "fit-page";
    if (handleImageLoadRafIdRef.current) cancelAnimationFrame(handleImageLoadRafIdRef.current);
    handleImageLoadRafIdRef.current = requestAnimationFrame(() => {
      handleImageLoadRafIdRef.current = 0;
      onZoomChange(pct);
    });
  }, [computeFitScale, clampToPct, onZoomChange]);
  React.useEffect(() => {
    if (!loaded) return;
    const container = containerRef.current;
    if (!container) return;
    let rafId = 0;
    const obs = new ResizeObserver(() => {
      if (rafId) cancelAnimationFrame(rafId);
      rafId = requestAnimationFrame(() => {
        rafId = 0;
        const mode = lastFitModeRef.current;
        if (mode === "manual") return;
        const scale = computeFitScale(mode);
        if (scale == null) return;
        onZoomChange(clampToPct(scale));
      });
    });
    obs.observe(container);
    return () => {
      obs.disconnect();
      if (rafId) cancelAnimationFrame(rafId);
    };
  }, [loaded, computeFitScale, clampToPct, onZoomChange]);
  const handleDoubleClick = React.useCallback(() => {
    if (!loaded) return;
    const fitScale = computeFitScale("fit-page");
    if (fitScale == null) return;
    const fitPct = clampToPct(fitScale);
    const atFit = Math.abs(zoom - fitPct) < 5;
    const targetPct = atFit ? 100 : fitPct;
    lastFitModeRef.current = atFit ? "manual" : "fit-page";
    onZoomChange(targetPct);
  }, [loaded, zoom, computeFitScale, clampToPct, onZoomChange]);
  React.useEffect(() => {
    const api = apiRef.current;
    if (!api || !loaded) return;
    const currentScale = api.state.scale;
    const targetScale = zoom / 100;
    if (Math.abs(currentScale - targetScale) < 5e-3) return;
    programmaticZoomRef.current = true;
    api.centerView(targetScale, 200);
    const t = setTimeout(() => {
      programmaticZoomRef.current = false;
    }, 280);
    return () => clearTimeout(t);
  }, [zoom, loaded]);
  React.useEffect(() => {
    if (!fitRequest || !loaded) return;
    const scale = computeFitScale(fitRequest.fit);
    if (scale == null) return;
    lastFitModeRef.current = fitRequest.fit;
    onZoomChange(clampToPct(scale));
  }, [fitRequest, loaded, computeFitScale, clampToPct, onZoomChange]);
  const handleTransformed = React.useCallback(
    (_ref, state) => {
      if (programmaticZoomRef.current) return;
      const nextZoom = Math.round(state.scale * 100);
      if (nextZoom !== zoom) {
        lastFitModeRef.current = "manual";
        onZoomChange(nextZoom);
      }
    },
    [zoom, onZoomChange]
  );
  return /* @__PURE__ */ jsx("div", { ref: containerRef, className: "w-full h-full overflow-hidden", onDoubleClick: handleDoubleClick, children: /* @__PURE__ */ jsx(
    TransformWrapper,
    {
      ref: apiRef,
      initialScale: 1,
      minScale: MIN_SCALE,
      maxScale: MAX_SCALE,
      centerOnInit: true,
      centerZoomedOut: true,
      limitToBounds: true,
      wheel: { step: 0.03 },
      doubleClick: { disabled: true },
      onTransform: handleTransformed,
      children: /* @__PURE__ */ jsx(TransformComponent, { wrapperClass: "!w-full !h-full", children: /* @__PURE__ */ jsx(
        "img",
        {
          ref: imgRef,
          src: file.url,
          alt: file.name,
          onLoad: handleImageLoad,
          draggable: false,
          className: "max-w-none max-h-none select-none",
          style: { pointerEvents: "none" }
        }
      ) })
    }
  ) });
};
ImageRenderer.displayName = "ImageRenderer";
const IMAGE_EXTS = /* @__PURE__ */ new Set(["jpg", "jpeg", "png", "gif", "webp", "svg", "avif", "bmp", "ico"]);
function canRenderImage(file) {
  var _a;
  if (file.mimeType.startsWith("image/")) return true;
  const ext = (_a = file.name.split(".").pop()) == null ? void 0 : _a.toLowerCase();
  return ext ? IMAGE_EXTS.has(ext) : false;
}
export {
  ImageRenderer,
  canRenderImage
};
