// @qijenchen/design-system/dist/components/Breadcrumb/breadcrumb.js
import { jsx, jsxs } from "react/jsx-runtime";
import * as React from "react";
import { Slot } from "@radix-ui/react-slot";
import { MoreHorizontal, ChevronRight } from "lucide-react";
import { cn } from "../../lib/utils.js";
import { ItemInlineActionButton } from "../../patterns/element-anatomy/item-anatomy.js";
import { Tooltip, TooltipTrigger, TooltipContent } from "../Tooltip/tooltip.js";
import { DropdownMenu, DropdownMenuTrigger, DropdownMenuContent, DropdownMenuItem } from "../DropdownMenu/dropdown-menu.js";
let sharedRO = null;
const sharedROCallbacks = /* @__PURE__ */ new WeakMap();
function getSharedRO() {
  if (sharedRO) return sharedRO;
  sharedRO = new ResizeObserver((entries) => {
    entries.forEach((e) => {
      const cb = sharedROCallbacks.get(e.target);
      if (cb) cb(e);
    });
  });
  return sharedRO;
}
function observeShared(el, cb) {
  const obs = getSharedRO();
  sharedROCallbacks.set(el, cb);
  obs.observe(el);
  return () => {
    sharedROCallbacks.delete(el);
    obs.unobserve(el);
  };
}
function TruncatedLabel({ children, fullText }) {
  const ref = React.useRef(null);
  const [isTruncated, setIsTruncated] = React.useState(false);
  React.useEffect(() => {
    const el = ref.current;
    if (!el) return;
    const check = () => setIsTruncated(el.scrollWidth > el.clientWidth);
    check();
    const raf = requestAnimationFrame(check);
    const t = setTimeout(check, 100);
    const cleanup = observeShared(el, check);
    return () => {
      cancelAnimationFrame(raf);
      clearTimeout(t);
      cleanup();
    };
  }, []);
  return /* @__PURE__ */ jsxs(Tooltip, { open: isTruncated ? void 0 : false, children: [
    /* @__PURE__ */ jsx(TooltipTrigger, { asChild: true, children: /* @__PURE__ */ jsx("span", { ref, className: "truncate min-w-0 block", children }) }),
    /* @__PURE__ */ jsx(TooltipContent, { children: fullText ?? children })
  ] });
}
const BreadcrumbContext = React.createContext({ size: "md" });
const BREADCRUMB_TEXT_CLASS = {
  sm: "text-body",
  md: "text-body",
  lg: "text-body-lg"
};
const BREADCRUMB_ICON_SIZE = {
  sm: 16,
  md: 16,
  lg: 20
};
const Breadcrumb = React.forwardRef(({ ...props }, ref) => /* @__PURE__ */ jsx("nav", { ref, "aria-label": "Breadcrumb", ...props }));
Breadcrumb.displayName = "Breadcrumb";
const BreadcrumbList = React.forwardRef(
  ({ className, size = "md", items, maxItems = 4, itemsBeforeCollapse = 1, itemsAfterCollapse = 1, children, ...props }, ref) => {
    const ctxValue = React.useMemo(() => ({ size }), [size]);
    const declarativeContent = React.useMemo(() => {
      if (!items) return null;
      const renderItem = (spec, role) => /* @__PURE__ */ jsx(BreadcrumbItem, { role, children: role === "current" ? /* @__PURE__ */ jsx(BreadcrumbPage, { startIcon: spec.startIcon, children: /* @__PURE__ */ jsx(TruncatedLabel, { fullText: typeof spec.label === "string" ? spec.label : void 0, children: spec.label }) }) : /* @__PURE__ */ jsx(BreadcrumbLink, { href: spec.href, asChild: spec.asChild, startIcon: spec.startIcon, children: /* @__PURE__ */ jsx(TruncatedLabel, { fullText: typeof spec.label === "string" ? spec.label : void 0, children: spec.label }) }) }, `${role}-${typeof spec.label === "string" ? spec.label : Math.random()}`);
      const shouldCollapse = items.length > maxItems;
      const beforeN = Math.max(0, itemsBeforeCollapse);
      const afterN = Math.max(1, itemsAfterCollapse);
      let visible;
      if (!shouldCollapse) {
        visible = items.map((spec, i) => ({
          spec,
          role: i === 0 ? "root" : i === items.length - 1 ? "current" : "middle"
        }));
      } else {
        const before = items.slice(0, beforeN).map((spec, i) => ({
          spec,
          role: i === 0 ? "root" : "middle"
        }));
        const collapsed = items.slice(beforeN, items.length - afterN);
        const after = items.slice(items.length - afterN).map((spec, i, arr) => ({
          spec,
          role: i === arr.length - 1 ? "current" : "middle"
        }));
        visible = [...before, { ellipsisOf: collapsed }, ...after];
      }
      const rendered = [];
      visible.forEach((entry, i) => {
        if (i > 0) rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, `sep-${i}`));
        if ("ellipsisOf" in entry) {
          rendered.push(
            /* @__PURE__ */ jsx(BreadcrumbItem, { role: "ellipsis", children: /* @__PURE__ */ jsxs(DropdownMenu, { children: [
              /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsx(BreadcrumbEllipsis, {}) }),
              /* @__PURE__ */ jsx(DropdownMenuContent, { align: "start", children: entry.ellipsisOf.map((s, j) => /* @__PURE__ */ jsx(DropdownMenuItem, { asChild: !!s.href, children: s.href ? /* @__PURE__ */ jsx("a", { href: s.href, children: s.label }) : /* @__PURE__ */ jsx("span", { children: s.label }) }, j)) })
            ] }) }, "ellipsis")
          );
        } else {
          rendered.push(renderItem(entry.spec, entry.role));
        }
      });
      return rendered;
    }, [items, maxItems, itemsBeforeCollapse, itemsAfterCollapse]);
    const compositionalContent = React.useMemo(() => {
      if (items) return null;
      const childArr = React.Children.toArray(children);
      const itemChildren = childArr.filter(
        (c) => {
          var _a;
          return React.isValidElement(c) && (c.type === BreadcrumbItem || ((_a = c.type) == null ? void 0 : _a.displayName) === "BreadcrumbItem");
        }
      );
      if (itemChildren.length === 0) return children;
      const total = itemChildren.length;
      const cloneWithRole = (item, idx, role) => React.cloneElement(item, { role: item.props.role ?? role, key: `bc-${role}-${idx}` });
      const shouldCollapse = total > maxItems;
      const beforeN = Math.max(0, itemsBeforeCollapse);
      const afterN = Math.max(1, itemsAfterCollapse);
      const rendered = [];
      if (!shouldCollapse) {
        itemChildren.forEach((item, i) => {
          if (i > 0) rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, `sep-${i}`));
          const role = i === 0 ? "root" : i === total - 1 ? "current" : "middle";
          rendered.push(cloneWithRole(item, i, role));
        });
      } else {
        const beforeItems = itemChildren.slice(0, beforeN);
        const collapsedItems = itemChildren.slice(beforeN, total - afterN);
        const afterItems = itemChildren.slice(total - afterN);
        beforeItems.forEach((item, i) => {
          if (i > 0) rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, `sep-bef-${i}`));
          const role = i === 0 ? "root" : "middle";
          rendered.push(cloneWithRole(item, i, role));
        });
        if (rendered.length > 0) rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, "sep-ellipsis-before"));
        rendered.push(
          /* @__PURE__ */ jsx(BreadcrumbItem, { role: "ellipsis", children: /* @__PURE__ */ jsxs(DropdownMenu, { children: [
            /* @__PURE__ */ jsx(DropdownMenuTrigger, { asChild: true, children: /* @__PURE__ */ jsx(BreadcrumbEllipsis, {}) }),
            /* @__PURE__ */ jsx(DropdownMenuContent, { align: "start", children: collapsedItems.map((item, j) => {
              const innerChildren = item.props.children;
              let href;
              let label = innerChildren;
              React.Children.forEach(innerChildren, (c) => {
                if (React.isValidElement(c)) {
                  if (c.props.href) href = c.props.href;
                  if (c.props.children != null) label = c.props.children;
                }
              });
              return href ? /* @__PURE__ */ jsx(DropdownMenuItem, { asChild: true, children: /* @__PURE__ */ jsx("a", { href, children: label }) }, `collapsed-${j}`) : /* @__PURE__ */ jsx(DropdownMenuItem, { children: label }, `collapsed-${j}`);
            }) })
          ] }) }, "bc-ellipsis")
        );
        rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, "sep-ellipsis-after"));
        afterItems.forEach((item, i) => {
          if (i > 0) rendered.push(/* @__PURE__ */ jsx(BreadcrumbSeparator, {}, `sep-aft-${i}`));
          const role = i === afterItems.length - 1 ? "current" : "middle";
          rendered.push(cloneWithRole(item, i, role));
        });
      }
      return rendered;
    }, [items, children, maxItems, itemsBeforeCollapse, itemsAfterCollapse]);
    return /* @__PURE__ */ jsx(BreadcrumbContext.Provider, { value: ctxValue, children: /* @__PURE__ */ jsx(
      "ol",
      {
        ref,
        className: cn(
          "flex flex-nowrap items-center gap-1 text-fg-secondary leading-compact min-w-0",
          BREADCRUMB_TEXT_CLASS[size],
          className
        ),
        ...props,
        children: items ? declarativeContent : compositionalContent
      }
    ) });
  }
);
BreadcrumbList.displayName = "BreadcrumbList";
const BreadcrumbItem = React.forwardRef(
  ({ className, role, style, ...props }, ref) => {
    const shrinkStyle = role === "root" ? { flexShrink: 3, minWidth: "1.5rem" } : role === "middle" ? { flexShrink: 2, minWidth: "1.5rem" } : role === "current" ? { flexShrink: 1, minWidth: "1.5rem" } : role === "ellipsis" ? { flexShrink: 0 } : {};
    return /* @__PURE__ */ jsx(
      "li",
      {
        ref,
        "data-bc-role": role,
        className: cn("inline-flex items-center min-w-0", className),
        style: { ...shrinkStyle, ...style },
        ...props
      }
    );
  }
);
BreadcrumbItem.displayName = "BreadcrumbItem";
const BreadcrumbLink = React.forwardRef(
  ({ asChild, className, children, startIcon: StartIcon, ...props }, ref) => {
    const { size } = React.useContext(BreadcrumbContext);
    const wrappedChildren = typeof children === "string" ? /* @__PURE__ */ jsx(TruncatedLabel, { fullText: children, children }) : children;
    const sharedClassName = cn(
      "inline-flex items-center gap-2",
      "min-w-0 max-w-full",
      "text-fg-secondary",
      "hover:text-primary-hover",
      "transition-colors duration-150",
      "focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-1",
      "rounded-md",
      className
    );
    if (asChild) {
      return /* @__PURE__ */ jsx(Slot, { ref, className: sharedClassName, ...props, children: wrappedChildren });
    }
    return /* @__PURE__ */ jsxs("a", { ref, className: sharedClassName, ...props, children: [
      StartIcon && /* @__PURE__ */ jsx(StartIcon, { size: BREADCRUMB_ICON_SIZE[size], "aria-hidden": true, className: "shrink-0" }),
      wrappedChildren
    ] });
  }
);
BreadcrumbLink.displayName = "BreadcrumbLink";
const BreadcrumbPage = React.forwardRef(
  ({ className, children, startIcon: StartIcon, ...props }, ref) => {
    const { size } = React.useContext(BreadcrumbContext);
    const wrappedChildren = typeof children === "string" ? /* @__PURE__ */ jsx(TruncatedLabel, { fullText: children, children }) : children;
    return /* @__PURE__ */ jsxs(
      "span",
      {
        ref,
        role: "link",
        "aria-disabled": "true",
        "aria-current": "page",
        className: cn("inline-flex items-center gap-2 min-w-0 max-w-full text-foreground", className),
        ...props,
        children: [
          StartIcon && /* @__PURE__ */ jsx(StartIcon, { size: BREADCRUMB_ICON_SIZE[size], "aria-hidden": true, className: "shrink-0" }),
          wrappedChildren
        ]
      }
    );
  }
);
BreadcrumbPage.displayName = "BreadcrumbPage";
const BreadcrumbSeparator = React.forwardRef(
  ({ children, className, ...props }, ref) => {
    const { size } = React.useContext(BreadcrumbContext);
    return /* @__PURE__ */ jsx(
      "li",
      {
        ref,
        role: "presentation",
        "aria-hidden": "true",
        className: cn("inline-flex items-center text-fg-muted shrink-0", className),
        ...props,
        children: children ?? /* @__PURE__ */ jsx(ChevronRight, { size: BREADCRUMB_ICON_SIZE[size], "aria-hidden": true })
      }
    );
  }
);
BreadcrumbSeparator.displayName = "BreadcrumbSeparator";
const BreadcrumbEllipsis = React.forwardRef(
  ({ "aria-label": ariaLabel = "顯示折疊路徑", ...props }, ref) => {
    return /* @__PURE__ */ jsx(
      ItemInlineActionButton,
      {
        ref,
        icon: MoreHorizontal,
        size: "md",
        "aria-label": ariaLabel,
        overlayTrigger: true,
        ...props
      }
    );
  }
);
BreadcrumbEllipsis.displayName = "BreadcrumbEllipsis";
const breadcrumbMeta = {
  component: "Breadcrumb",
  family: null,
  // non-family composite / overlay / layout
  variants: {},
  sizes: {},
  states: ["default", "hover", "active", "focus-visible", "disabled"],
  tokens: {
    bg: [],
    fg: ["text-fg-muted", "text-fg-secondary", "text-foreground"],
    ring: ["ring-ring"]
  }
};
export {
  Breadcrumb,
  BreadcrumbEllipsis,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
  breadcrumbMeta
};
