// @qijenchen/design-system/dist/patterns/horizontal-overflow/horizontal-overflow.js
import { jsx } from "react/jsx-runtime";
import * as React from "react";
import { ChevronDown, ChevronLeft, ChevronRight } from "lucide-react";
import { cn } from "../../lib/utils.js";
import { Button } from "../../components/Button/button.js";
const FADE_WIDTH = 16;
const ARROW_BUTTON_WIDTH = 32;
const SCROLL_PAGE_RATIO = 0.8;
function buildFadeMask({
  canScroll,
  atStart,
  atEnd,
  reserveArrowWidth = 0
}) {
  if (!canScroll) return void 0;
  const stops = [];
  if (!atStart) {
    if (reserveArrowWidth > 0) {
      stops.push("transparent 0");
      stops.push(`transparent ${reserveArrowWidth}px`);
      stops.push(`black ${reserveArrowWidth + FADE_WIDTH}px`);
    } else {
      stops.push("transparent 0");
      stops.push(`black ${FADE_WIDTH}px`);
    }
  } else {
    stops.push("black 0");
  }
  if (!atEnd) {
    if (reserveArrowWidth > 0) {
      stops.push(`black calc(100% - ${reserveArrowWidth + FADE_WIDTH}px)`);
      stops.push(`transparent calc(100% - ${reserveArrowWidth}px)`);
      stops.push("transparent 100%");
    } else {
      stops.push(`black calc(100% - ${FADE_WIDTH}px)`);
      stops.push("transparent 100%");
    }
  } else {
    stops.push("black 100%");
  }
  return `linear-gradient(to right, ${stops.join(", ")})`;
}
function useScrollByPage(scrollRef) {
  return React.useCallback(
    (direction) => {
      const el = scrollRef.current;
      if (!el) return;
      el.scrollBy({
        left: el.clientWidth * SCROLL_PAGE_RATIO * (direction === "left" ? -1 : 1),
        behavior: "smooth"
      });
    },
    [scrollRef]
  );
}
const OverflowScrollArrow = ({
  direction,
  onClick,
  "aria-label": ariaLabel
}) => {
  const defaultLabel = direction === "left" ? "向左捲動" : "向右捲動";
  return /* @__PURE__ */ jsx(
    "div",
    {
      className: cn(
        "absolute top-0 bottom-0 flex items-center pointer-events-none z-10",
        direction === "left" ? "left-0" : "right-0"
      ),
      children: /* @__PURE__ */ jsx("div", { className: "pointer-events-auto", children: /* @__PURE__ */ jsx(
        Button,
        {
          variant: "text",
          size: "sm",
          iconOnly: true,
          startIcon: direction === "left" ? ChevronLeft : ChevronRight,
          "aria-label": ariaLabel ?? defaultLabel,
          onClick
        }
      ) })
    }
  );
};
const OverflowMenuTriggerButton = React.forwardRef(({ "aria-label": ariaLabel, ...props }, ref) => /* @__PURE__ */ jsx(
  Button,
  {
    ref,
    variant: "text",
    size: "sm",
    iconOnly: true,
    startIcon: ChevronDown,
    "aria-label": ariaLabel,
    ...props
  }
));
OverflowMenuTriggerButton.displayName = "OverflowMenuTriggerButton";
export {
  ARROW_BUTTON_WIDTH,
  FADE_WIDTH,
  OverflowMenuTriggerButton,
  OverflowScrollArrow,
  SCROLL_PAGE_RATIO,
  buildFadeMask,
  useScrollByPage
};
