// @qijenchen/design-system/dist/components/Dialog/dialog.js
import { jsx, jsxs } from "react/jsx-runtime";
import * as React from "react";
import * as DialogPrimitive from "@radix-ui/react-dialog";
import { X } from "lucide-react";
import { cn } from "../../lib/utils.js";
import { Button } from "../Button/button.js";
import { SurfaceFooter, SurfaceHeader } from "../../patterns/overlay-surface/overlay-surface.js";
import { ScrollArea } from "../ScrollArea/scroll-area.js";
const Dialog = DialogPrimitive.Root;
const DialogTrigger = DialogPrimitive.Trigger;
const DialogPortal = DialogPrimitive.Portal;
const DialogClose = DialogPrimitive.Close;
const DIALOG_INSET_VAR = "var(--layout-space-bottom)";
const DialogOverlay = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Overlay,
  {
    ref,
    className: cn(
      "fixed inset-0 z-50 bg-overlay",
      "data-[state=open]:animate-in data-[state=closed]:animate-out motion-reduce:animate-none",
      "data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0",
      className
    ),
    ...props
  }
));
DialogOverlay.displayName = DialogPrimitive.Overlay.displayName;
const DialogContent = React.forwardRef(({ className, maxWidth = "512px", autoHeight, children, style, ...props }, ref) => {
  const insetCalc = `${DIALOG_INSET_VAR} * 2`;
  const viewportH = `calc(100vh - ${insetCalc})`;
  const maxWidthCss = typeof maxWidth === "number" ? `${maxWidth}px` : maxWidth;
  const heightStyle = autoHeight ? { maxHeight: viewportH } : { height: viewportH };
  const handleOpenAutoFocus = (e) => {
    e.preventDefault();
    const content = e.currentTarget;
    const firstBodyTarget = content.querySelector(
      "[data-dialog-body] input:not([disabled]),[data-dialog-body] textarea:not([disabled]),[data-dialog-body] select:not([disabled]),[data-dialog-body] button:not([disabled]):not([data-dismiss])"
    );
    const firstFooterButton = content.querySelector(
      "[data-dialog-footer] button:not([disabled]):not([data-dismiss])"
    );
    (firstBodyTarget ?? firstFooterButton ?? content).focus({ preventScroll: true });
  };
  return /* @__PURE__ */ jsxs(DialogPortal, { children: [
    /* @__PURE__ */ jsx(DialogOverlay, {}),
    /* @__PURE__ */ jsx(
      DialogPrimitive.Content,
      {
        ref,
        onOpenAutoFocus: handleOpenAutoFocus,
        className: cn(
          "fixed left-1/2 top-1/2 z-50 w-full -translate-x-1/2 -translate-y-1/2",
          "flex flex-col bg-surface-raised rounded-lg border border-border",
          "data-[state=open]:animate-in data-[state=closed]:animate-out motion-reduce:animate-none",
          "data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0",
          "data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95",
          "data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%]",
          "data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%]",
          className
        ),
        style: {
          boxShadow: "var(--elevation-200)",
          maxWidth: `min(${maxWidthCss}, calc(100vw - ${insetCalc}))`,
          ...heightStyle,
          ...style
        },
        ...props,
        children
      }
    )
  ] });
});
DialogContent.displayName = DialogPrimitive.Content.displayName;
const DialogHeader = React.forwardRef(({ className, children, ...props }, ref) => (
  // 2026-05-18:className 不再硬加 justify-between(冗餘:row 1 是 flex items-center gap-2,
  // 第一 child flex-1 grow 自然 push close X 靠右,跟 justify-between 同視覺)。
  // 並且 column mode(tabsSlot 提供)justify-between 會把 row 1 / row 2 上下推開 = 破裂。
  // tabsSlot via `...props` spread 自動 forward(type 來自 SurfaceHeaderProps)。
  /* @__PURE__ */ jsxs(
    SurfaceHeader,
    {
      ref,
      className,
      ...props,
      children: [
        /* @__PURE__ */ jsx("div", { className: "flex-1 min-w-0", children }),
        /* @__PURE__ */ jsx(DialogPrimitive.Close, { asChild: true, children: /* @__PURE__ */ jsx(Button, { "data-dismiss": true, iconOnly: true, dismiss: true, size: "sm", startIcon: X, "aria-label": "關閉" }) })
      ]
    }
  )
));
DialogHeader.displayName = "DialogHeader";
const DialogBody = React.forwardRef(({ className, children, ...props }, ref) => /* @__PURE__ */ jsx(ScrollArea, { ref, "data-dialog-body": true, className: "flex-1 min-h-0", ...props, children: /* @__PURE__ */ jsx(
  "div",
  {
    className: cn(
      "px-[var(--layout-space-loose)] pt-[var(--layout-space-tight)] pb-[var(--layout-space-bottom)]",
      className
    ),
    children
  }
) }));
DialogBody.displayName = "DialogBody";
const DialogFooter = React.forwardRef(({ ...props }, ref) => /* @__PURE__ */ jsx(SurfaceFooter, { ref, "data-dialog-footer": true, ...props }));
DialogFooter.displayName = "DialogFooter";
const DialogTitle = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Title,
  {
    ref,
    className: cn("text-body-lg font-medium truncate", className),
    ...props
  }
));
DialogTitle.displayName = DialogPrimitive.Title.displayName;
const DialogDescription = React.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ jsx(
  DialogPrimitive.Description,
  {
    ref,
    className: cn("mt-[var(--item-gap-label-desc-reading-lg)] text-body text-fg-secondary", className),
    ...props
  }
));
DialogDescription.displayName = DialogPrimitive.Description.displayName;
const dialogMeta = {
  component: "Dialog",
  family: null,
  // non-family composite / overlay / layout
  variants: {},
  sizes: {},
  states: ["default", "hover", "active", "focus-visible", "disabled"],
  tokens: {
    bg: ["bg-surface-raised"],
    fg: ["text-fg-secondary"],
    ring: []
  }
};
export {
  Dialog,
  DialogBody,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogOverlay,
  DialogPortal,
  DialogTitle,
  DialogTrigger,
  dialogMeta
};
