// @qijenchen/design-system/dist/lib/drag-visual.js
const dragSourceClass = "opacity-disabled";
function dragSourceStyle(isDragging) {
  return isDragging ? { opacity: "var(--opacity-disabled)" } : {};
}
const dropIndicatorRow = {
  before: "absolute top-0 left-0 right-0 h-0.5 bg-primary z-10 pointer-events-none",
  after: "absolute bottom-0 left-0 right-0 h-0.5 bg-primary z-10 pointer-events-none"
};
const dropIndicatorColumn = {
  // Absolute div 變體(consumer render <div>)
  before: "absolute top-0 bottom-0 left-[-1px] w-0.5 bg-primary z-10 pointer-events-none",
  after: "absolute top-0 bottom-0 right-[-1px] w-0.5 bg-primary z-10 pointer-events-none",
  // Pseudo-element 變體(用 Tailwind before:/after:)
  pseudoBefore: 'before:absolute before:top-0 before:bottom-0 before:left-[-1px] before:w-0.5 before:bg-primary before:z-10 before:pointer-events-none before:content-[""]',
  pseudoAfter: 'after:absolute after:top-0 after:bottom-0 after:right-[-1px] after:w-0.5 after:bg-primary after:z-10 after:pointer-events-none after:content-[""]'
};
const dropIndicatorInside = "bg-primary-subtle";
const dragActiveCursor = "cursor-grabbing";
const dragHandleCursorClass = "cursor-grab";
const snapToCursorModifier = ({ transform, activatorEvent, draggingNodeRect }) => {
  if (!activatorEvent || !draggingNodeRect) return transform;
  if (!("clientX" in activatorEvent) || !("clientY" in activatorEvent)) return transform;
  const ev = activatorEvent;
  if (typeof ev.clientX !== "number" || typeof ev.clientY !== "number") return transform;
  const offsetX = ev.clientX - draggingNodeRect.left;
  const offsetY = ev.clientY - draggingNodeRect.top;
  return {
    ...transform,
    x: transform.x + offsetX,
    y: transform.y + offsetY
  };
};
function isReorderNoop(activeIdx, overIdx, side) {
  if (activeIdx === overIdx) return true;
  if (side === "after" && overIdx + 1 === activeIdx) return true;
  if (side === "before" && overIdx - 1 === activeIdx) return true;
  return false;
}
function reconstructFullRowGhost(rowId, tableRoot) {
  if (!tableRoot) return null;
  const escaped = rowId.replace(/(["\\])/g, "\\$1");
  const allRows = Array.from(
    tableRoot.querySelectorAll(
      `[role="row"][data-sortable-row-id="${escaped}"]`
    )
  );
  if (allRows.length === 0) return null;
  if (allRows.length === 1) {
    const sourceEl = allRows[0];
    const clone = sourceEl.cloneNode(true);
    sanitizeGhostClone(clone, sourceEl.offsetWidth);
    return { html: clone.outerHTML, width: sourceEl.offsetWidth };
  }
  const primaryRow = allRows.find((r) => r.dataset.rowDragSource === "true") ?? allRows[0];
  const cellsHtml = [];
  let totalWidth = 0;
  for (const regionRow of allRows) {
    totalWidth += regionRow.offsetWidth;
    for (const cell of Array.from(
      regionRow.querySelectorAll('[role="cell"],[role="gridcell"]')
    )) {
      const cellClone = cell.cloneNode(true);
      cellClone.querySelectorAll("[data-drag-handle-portal]").forEach((n) => n.remove());
      cellClone.style.transform = "none";
      cellsHtml.push(cellClone.outerHTML);
    }
  }
  const inlineStyle = `display:flex; width:${totalWidth}px; transform:none; position:static; opacity:1; transition:none;`;
  const html = `<div role="row" class="${primaryRow.className}" style="${inlineStyle}">${cellsHtml.join("")}</div>`;
  return { html, width: totalWidth };
}
function sanitizeGhostClone(el, width) {
  el.style.position = "static";
  el.style.transform = "none";
  el.style.transition = "none";
  el.style.opacity = "1";
  el.style.zIndex = "";
  el.style.width = `${width}px`;
  el.removeAttribute("data-row-index");
  el.removeAttribute("aria-rowindex");
  el.removeAttribute("data-hovered");
  el.querySelectorAll("[data-drag-handle-portal]").forEach((n) => n.remove());
}
export {
  dragActiveCursor,
  dragHandleCursorClass,
  dragSourceClass,
  dragSourceStyle,
  dropIndicatorColumn,
  dropIndicatorInside,
  dropIndicatorRow,
  isReorderNoop,
  reconstructFullRowGhost,
  snapToCursorModifier
};
